README.md
### Synopsis

This package provides efficient integer interval sets.

### Description

> Persistent... is it trees?

Yes, Radix trees. Trees are balanced by prefix bits, so we have fast
merge operations, such as union, intersection and difference.  Chris
Okasaki and Andrew Gill [shows][int-maps] that Patricia tree based
integer maps might be order of magnitude faster than Red-Black tree
counterparts on this operations. The same apply to integer sets, we
just have keys, but don't have values.

> That does mean the "dense"?

That means we keep suffixes in bitmaps and we might pack, say 10,
integers which lies close together in one bitmap. This optimization
doesn't affect execution times for sparse sets, but makes dense sets
much more memory efficient — near 10-50 times less space usage
depending on machine word size and the actual density of the
set. Basically, this let us be 3-4 times less memory efficient
comparing with arrays of tightly packed bits, but see...

> How suffix compaction is performed?

There are exist a pretty simple algorithm used in memory allocators
called ["buddy memory allocator"][buddy-alloc]. In a nutshell, we have
a big block which is splitted by half when we remove from one of the
half, and merge then back when we insert. It's somewhat inverse to the
ordinary tree approach — in buddy tree we hold more information about
elements that it _doesn't_ contain, while in prefix tree we hold more
information about elements that it _does_ contain. It's easy to guess
that we should do with it — take the two structures then fuse them
into one to produce a new structure which perform _better_.

Indeed, the key idea in the design is right here — we switch forth and
back between representations per subtree basis. We intersperse
different representations in different tree branches. It's like
chameleon:

* If the some subset is _sparse_, we just keep a radix tree with
  bitmaps at leafs.

* If the some subset becomes _full_ we turn it into block. If some
  buddy block appears, we join the buddy blocks into one. And so forth.

That is, we just get a structure that dynamically choose the optimal
representation depending on _density_ of set. Moreover in best case
this lead to huge space savings:


``` haskell
> ppStats (fromList [0..123456])
```

gives:

```
Bin count: 6
Tip count: 1
Fin count: 6
Size in bytes: 408
Saved space over dense set:  123072
Saved space over bytestring: 11879
Percent saved over dense set:  99.6695821185617%
Percent saved over bytestring: 96.67941727028567%
```

The `ppStats` is not an exposed function but you can play with it
using `cabal-dev ghci`.

I don't know if it is an old idea, but this works just fine.

> So when this data structure is good choice?

In many situation. It might be used as persistent and compact
replacement for bool arrays or Data.IntSet with the following advantages:

* Purity is extremely useful in multithreaded settings —
  we could keep a set in a mutable transactional variable or an IORef
  and atomically update/modify the set. So it could be used as
  replacement for [TArray Int Bool][tarray] as well.
* By merging intervals together we achieve compactness. In best case
  some of main operations will take O(1)time and space, so if you need
  interval set it's here.
* Fast serizalization: if you are need conversion to/from bytestrings.
  Because of bitmaps it's possible to do this conversion _extremely_ fast.

> How this implementation relate to containers version?

Heavely based. Essentially we just add the buddy interval compaction,
but it turns out that some operations becomes more complicated and
requires much more effort to implement — in order to maintain the all
tree invariants we need to take into account more cases. This is the
reason why some operations are not implemented yet (e.g. lack of
views), but I hope I'll fix it with the time.

### Documentation

For documentation see haddock generated documentation.

### Build Status

[![Build Status][status-img]][status-link]

### Maintainer

This library is written and maintained by Sam T. <pxqr.sta@gmail.com>

Feel free to report bugs and suggestions via
[github issue tracker][issues] or the mail.

[status-img]:  https://travis-ci.org/pxqr/intset.png
[status-link]: https://travis-ci.org/pxqr/intset
[issues]:      https://github.com/pxqr/intset/issues/new

[int-maps]:    http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.37.5452
[buddy-alloc]: http://en.wikipedia.org/wiki/Buddy_memory_allocation
[tarray]:      http://hackage.haskell.org/packages/archive/stm/2.2.0.1/doc/html/Control-Concurrent-STM-TArray.html


ChangeLog
2013-12-07  Sam Truzjan  <pxqr.sta@gmail.com>

	* 0.1.1.0: Make Show instance compatible with containers package

2013-10-04  Sam Truzjan  <pxqr.sta@gmail.com>

	* 0.1.0.3: FIX: Add changelog to tarball.

2013-10-04  Sam Truzjan  <pxqr.sta@gmail.com>

	* 0.1.0.2: Move release notes to changelog file.

2013-08-12  Sam Truzjan  <pxqr.sta@gmail.com>

	* 0.1.0.1: Fix build failure on 32 bit arch.

2013-06-08  Sam Truzjan  <pxqr.sta@gmail.com>

	* 0.1.0.0: Initial version.


tests/Main.hs
{-# OPTIONS -fno-warn-orphans #-}
module Main (main) where

import Control.Applicative hiding (empty)
import Test.QuickCheck hiding ((.&.))
import Test.Framework
import Test.Framework.Providers.QuickCheck2

import Data.List as L (sort, nub, map, filter, minimum, intersect)
import Data.IntervalSet as S
import Data.IntervalSet.ByteString as S
import Data.Monoid



instance Arbitrary IntSet where
  arbitrary = oneof [fromList <$> arbitrary, buddy <$> arbitrary]
    where
      buddy :: [Int] -> IntSet
      buddy = fromList . concatMap mk
        where
          mk i = [i * 64 .. i * 64 + 64]

  shrink (Bin _ _ l r) = [l, r]
  shrink (Fin p m)     = [splitFin p m]
  shrink  _            = []

prop_empty :: [Int] -> Bool
prop_empty xs = (not . (`member` empty)) `all` xs

prop_singleton :: Int -> [Int] -> Bool
prop_singleton e = all check
  where
    check x |   x == e  = member    x (singleton e)
            | otherwise = notMember x (singleton e)

prop_insertLookup :: IntSet -> Int -> Bool
prop_insertLookup s i = member i (insert i s )

prop_unionLookup :: [Int] -> [Int] -> Bool
prop_unionLookup a b = all (`member` u) a && all (`member` u) b
  where
    u = fromList a `union` fromList b

prop_sorted :: IntSet -> Bool
prop_sorted xs = toList xs == L.nub (L.sort (toList xs))

prop_valid :: IntSet -> Bool
prop_valid = isValid

prop_unionSize :: IntSet -> IntSet -> Bool
prop_unionSize a b = size u >= size a
                  && size u >= size b
                  && size u <= size a + size b
  where
    u = a `union` b

prop_unionComm :: IntSet -> IntSet -> Bool
prop_unionComm a b = a <> b == b <> a

prop_unionAssoc :: IntSet -> IntSet -> IntSet -> Bool
prop_unionAssoc a b c = a <> (b <> c) == (a <> b) <> c

prop_unionLeftId :: IntSet -> Bool
prop_unionLeftId a = mempty <> a == a

prop_unionRightId :: IntSet -> Bool
prop_unionRightId a = mempty <> a == a

prop_unionTop :: IntSet -> Bool
prop_unionTop a = a <> a == a

prop_unionIdemp :: IntSet -> IntSet -> Bool
prop_unionIdemp a b = ((a <> b) <> b) == a <> b

prop_intersectionSize :: IntSet -> IntSet -> Bool
prop_intersectionSize a b = size i <= size a && size i <= size b
  where
    i = intersection a b

prop_intersection :: [Int] -> [Int] -> Bool
prop_intersection a b =
  fromList a `intersection` fromList b
  == fromList (a `L.intersect` b)

prop_intersectComm :: IntSet -> IntSet -> Bool
prop_intersectComm a b = (a `intersection` b) == (b `intersection` a)

prop_intersectAssoc :: IntSet -> IntSet -> IntSet -> Bool
prop_intersectAssoc a b c =
  ((a `intersection` b) `intersection` c)
  ==  (a `intersection` (b `intersection` c))

prop_intersectLeft :: IntSet -> Bool
prop_intersectLeft a = intersection empty a == empty

prop_intersectRight :: IntSet -> Bool
prop_intersectRight a = intersection a empty == empty

prop_intersectBot :: IntSet -> Bool
prop_intersectBot a = (a `intersection` a) == a

prop_intersectIdemp :: IntSet -> IntSet -> Bool
prop_intersectIdemp a b = ((a `intersection` b) `intersection` b)
                          == intersection a b

prop_showRead :: IntSet -> Bool
prop_showRead a = read (show a) == a

prop_eq :: [Int] -> [Int] -> Bool
prop_eq a b
    | a' == b' = fromList a' == fromList b'
    | a' /= b' = fromList a' /= fromList b'
    | otherwise = error "prop_eq: impossible"
  where
    a' = nub (sort a)
    b' = nub (sort b)

prop_eqRefl :: IntSet -> Bool
prop_eqRefl a = a == a

prop_eqSym :: IntSet -> IntSet -> Bool
prop_eqSym a b = (a == b) == (b == a)

prop_eqTrans :: IntSet -> IntSet -> IntSet -> Bool
prop_eqTrans a b c
  | (a == b) && (b == c) = a == c
  |    otherwise         = True

prop_size :: [Int] -> Bool
prop_size xs = length (nub (sort xs)) == size (fromList xs)

prop_insertDelete :: Int -> IntSet -> Bool
prop_insertDelete i = notMember i . delete i . insert i

prop_mapPresSize :: IntSet -> Bool
prop_mapPresSize s = size (S.map (*2) s) == size s

prop_mapLessSize :: IntSet -> Bool
prop_mapLessSize s = size (S.map (`div` 2) s) <= size s

prop_mapping :: [Int] -> Bool
prop_mapping xs = toList (S.map (*2) (fromList xs)) == L.map (*2) (nub (sort xs))

prop_filterSize :: IntSet -> Bool
prop_filterSize s = size (S.filter even s) <= size s

prop_filtering :: [Int] -> Bool
prop_filtering xs = S.filter even (fromList xs) == fromList (L.filter even xs)

prop_min :: [Int] -> Bool
prop_min [] = True
prop_min xs = findMin (fromList xs) == L.minimum xs

{-
prop_universeMember :: [Int] -> Bool
prop_universeMember = all (`member` universe)

prop_universeDelete :: Int -> Bool
prop_universeDelete i = notMember i (delete i universe)

prop_universeInsert :: Int -> Bool
prop_universeInsert i = insert i universe == universe

prop_universeNatNeg :: Bool
prop_universeNatNeg = naturals <> negatives == universe

prop_naturals :: [Int] -> Bool
prop_naturals = all check
  where
    check x |   x >= 0  = member x naturals
            | otherwise = True

prop_negatives :: [Int] -> Bool
prop_negatives = all check
  where
    check x |   x < 0   = member x negatives
            | otherwise = True
-}

prop_minInSet :: IntSet -> Bool
prop_minInSet s
  |  S.null s = True
  | otherwise = member (findMin s) s

prop_minIsTheLess :: IntSet -> Bool
prop_minIsTheLess s
  | S.null s  = True
  | otherwise = all (findMin s <=) (toList s)

prop_maxInSet :: IntSet -> Bool
prop_maxInSet s
  |  S.null s = True
  | otherwise = member (findMax s) s

prop_maxIsTheGreatest :: IntSet -> Bool
prop_maxIsTheGreatest s
  |  S.null s = True
  | otherwise = all (<= findMax s) (toList s)

prop_differenceMember :: IntSet -> IntSet -> Bool
prop_differenceMember a b = all (`notMember` difference a b) (toList b)

prop_differenceIntersection :: IntSet -> IntSet -> Bool
prop_differenceIntersection a b = (difference a b `intersection` b) == empty

prop_differenceSize :: IntSet -> IntSet -> Bool
prop_differenceSize a b = size (difference a b) <= size a


prop_differenceSubset :: IntSet -> IntSet -> Bool
prop_differenceSubset a b = (a `difference` b) `isSubsetOf` a


prop_differenceDeMorgan1 :: IntSet -> IntSet -> IntSet -> Bool
prop_differenceDeMorgan1 a b c = a - b * c == (a - b) + (a - c)

prop_differenceDeMorgan2 :: IntSet -> IntSet -> IntSet -> Bool
prop_differenceDeMorgan2 a b c = a - (b + c) == (a - b) * (a - c)

prop_differenceDistributive :: IntSet -> IntSet -> IntSet -> Bool
prop_differenceDistributive a b c = (a + b) - c == (a - c) + (b - c)

prop_splitPivot :: IntSet -> Key -> Bool
prop_splitPivot s k = all (< k) (toList lt) && all (k <) (toList gt)
  where
    (lt, gt) = split k s

prop_splitIntersect :: IntSet -> Key -> Bool
prop_splitIntersect s k = lt * gt == empty
  where
    (lt, gt) = split k s

prop_splitGT :: IntSet -> Key -> Bool
prop_splitGT s k = all (k <) (toList (splitGT k s))

prop_splitLT :: IntSet -> Key -> Bool
prop_splitLT s k = all (< k) (toList (splitLT k s))

prop_interval :: Int -> Int -> Bool
prop_interval a s = interval l r == fromList [l..r]
  where
    l = a
    r = l + min 10000 s

prop_cmp :: IntSet -> IntSet -> Bool
prop_cmp a b = compare a b == compare (toList a) (toList b)

prop_numInst :: Int -> Bool
prop_numInst i = fromIntegral i == singleton i

prop_deleteEmpty :: Int -> Bool
prop_deleteEmpty k = delete k empty == empty

prop_elems :: IntSet -> Bool
prop_elems s = toList s == elems s

prop_combine :: [IntSet] -> Bool
prop_combine xs = all check xs
  where
    check x = us <> x == us && is * x == is
    us = mconcat xs
    is = intersections xs

prop_sortIdemp :: [Int] -> Bool
prop_sortIdemp xs = let a = ssort xs in ssort a == a
  where
    ssort = toList . fromList

prop_bitmapEncode :: IntSet -> Bool
prop_bitmapEncode xs = fromByteString (toByteString xs') == xs'
  where -- we should restrict upper bound otherwise we might have out of memory
    xs' = splitGT (-1) $ splitLT 1000000 xs

prop_partition :: IntSet -> Bool
prop_partition s = fst (S.partition even s) == S.filter even s
               &&  snd (S.partition even s) == S.filter odd  s

prop_subsetSize :: IntSet -> IntSet -> Bool
prop_subsetSize a b
  | a `isSubsetOf` b = size a <= size b
  |     otherwise    = True

prop_supersetSize :: IntSet -> IntSet -> Bool
prop_supersetSize a b
  | a `isSupersetOf` b = size a >= size b
  |    otherwise       = True

prop_subsetSuperset :: IntSet -> IntSet -> Bool
prop_subsetSuperset a b = (a `isSubsetOf` b) || (b `isSubsetOf` a)
                          || not (S.null (symDiff a b))

prop_subsetIntersection :: IntSet -> IntSet -> Bool
prop_subsetIntersection a b = (i `isSubsetOf` a) && (i `isSubsetOf` b)
  where
    i = a * b

prop_subsetUnion :: IntSet -> IntSet -> Bool
prop_subsetUnion a b = (a `isSubsetOf` u) && (b `isSubsetOf` u)
  where
    u = a + b


prop_symDiffLeftNeutral :: IntSet -> Bool
prop_symDiffLeftNeutral a = symDiff empty a == a

prop_symDiffRightNeutral :: IntSet -> Bool
prop_symDiffRightNeutral a = symDiff a empty == a

prop_symDiffCommutative :: IntSet -> IntSet -> Bool
prop_symDiffCommutative a b = (a `symDiff` b) == b `symDiff` a

prop_symDiffAssociative :: IntSet -> IntSet -> IntSet -> Bool
prop_symDiffAssociative a b c = ((a `symDiff` b) `symDiff` c)
                             == (a `symDiff` (b `symDiff` c))

prop_symDiffDistr :: IntSet -> IntSet -> IntSet -> Bool
prop_symDiffDistr a b c = (a `intersection` (b  `symDiff` c))
                       == ((a `intersection` b) `symDiff` (a `intersection` c))

prop_symDiffUnion :: IntSet -> IntSet -> Bool
prop_symDiffUnion a b = (a `difference` b) `union` (b `difference` a)
                     == symDiff a b

prop_symDiffInter :: IntSet -> IntSet -> Bool
prop_symDiffInter a b = (a `union` b) `difference` (a `intersection` b)
  == symDiff a b

prop_symDiffSizeBound :: IntSet -> IntSet -> Bool
prop_symDiffSizeBound a b = size s <= size a + size b
  where
    s = symDiff a b

main :: IO ()
main = defaultMain
  [ testProperty "empty"                prop_empty
  , testProperty "singleton"            prop_singleton
  , testProperty "insertLookup"         prop_insertLookup
  , testProperty "insert delete"        prop_insertDelete
  , testProperty "compare"              prop_cmp
  , testProperty "interval"             prop_interval

  , testProperty "subset size"          prop_subsetSize
  , testProperty "superset size"        prop_supersetSize
  , testProperty "subset superset exclusion"  prop_subsetSuperset
  , testProperty "subset of intersection"     prop_subsetIntersection
  , testProperty "subset of union"            prop_subsetUnion

  , testProperty "bitmap_encode"        prop_bitmapEncode

  , testProperty "symmetric difference left neutral"  prop_symDiffLeftNeutral
  , testProperty "symmetric difference right neutral" prop_symDiffRightNeutral
  , testProperty "symmetric difference commutative"   prop_symDiffCommutative
  , testProperty "symmetric difference associative"   prop_symDiffAssociative
  , testProperty "symmetric difference distributive"  prop_symDiffDistr
  , testProperty "symmetric difference union"         prop_symDiffUnion
  , testProperty "symmetric difference intersection"  prop_symDiffInter
  , testProperty "symmetric difference size upper bound" prop_symDiffSizeBound


{-
  , testProperty "universe member"      prop_universeMember
  , testProperty "universe delete"      prop_universeDelete
  , testProperty "universe insert"      prop_universeInsert
  , testProperty "universe nat neg"     prop_universeNatNeg
  , testProperty "naturals"             prop_naturals
  , testProperty "negatives"            prop_negatives
-}

  , testProperty "size"                 prop_size
  , testProperty "sort"                 prop_sorted
  , testProperty "sort idempotent"      prop_sortIdemp


  , testProperty "read . show == id"    prop_showRead

  , testProperty "equality"              prop_eq
  , testProperty "equality reflexivity"  prop_eqRefl
  , testProperty "equality symmetry"     prop_eqSym
  , testProperty "equality transitivity" prop_eqTrans

  , testProperty "map preserve size"    prop_mapPresSize
  , testProperty "map not preserve siz" prop_mapLessSize
  , testProperty "mapping"              prop_mapping

  , testProperty "filter size"          prop_filterSize
  , testProperty "filtering"            prop_filtering

  , testProperty "union size"           prop_unionSize
  , testProperty "unionLookup"          prop_unionLookup
  , testProperty "union commutative"    prop_unionComm
  , testProperty "union associative"    prop_unionAssoc
  , testProperty "union left identity"  prop_unionLeftId
  , testProperty "union right identity" prop_unionRightId
  , testProperty "union top"            prop_unionTop
  , testProperty "union idemp"          prop_unionIdemp

  , testProperty "minimal in set"        prop_minInSet
  , testProperty "maximal in set"        prop_maxInSet
  , testProperty "minimal is the less"   prop_minIsTheLess
  , testProperty "maximal is the greatest" prop_maxIsTheGreatest


  , testProperty "intersection size"         prop_intersectionSize
  , testProperty "intersection lists"        prop_intersection
  , testProperty "intersection commutative"  prop_intersectComm
  , testProperty "intersection associative"  prop_intersectAssoc
  , testProperty "intersection idemp"        prop_intersectIdemp
  , testProperty "intersection left empty"   prop_intersectLeft
  , testProperty "intersection right empty"  prop_intersectRight
  , testProperty "intersection bot"          prop_intersectBot

  , testProperty "difference member"         prop_differenceMember
  , testProperty "difference intersection"   prop_differenceIntersection
  , testProperty "difference size"           prop_differenceSize
  , testProperty "difference de morgan1"     prop_differenceDeMorgan1
  , testProperty "difference de morgan2"     prop_differenceDeMorgan2
  , testProperty "difference distributive"   prop_differenceDistributive
  , testProperty "difference subset"         prop_differenceSubset

  , testProperty "split pivot"               prop_splitPivot
  , testProperty "split intersection"        prop_splitIntersect
  , testProperty "split greater than"        prop_splitGT
  , testProperty "split lesser  than"        prop_splitLT
  , testProperty "partition filter"          prop_partition

  , testProperty "min"                  prop_min
  , testProperty "valid"                prop_valid

    -- for coverage mostly
  , testProperty "delete from empty"    prop_deleteEmpty
  , testProperty "combine"              prop_combine
  , testProperty "num instance"         prop_numInst
  , testProperty "elems"                prop_elems
  ]


tests/Fusion.hs
module Main (main) where

import Data.IntervalSet as S
import System.Exit

-- should fuse to id
test :: [Int] -> [Int]
test x = toList (S.map id (fromList x))

main :: IO ()
main = do
  print $ test []
  exitSuccess

src/Data/IntervalSet.hs
-- |
--   Copyright   :  (c) Sam Truzjan 2013
--   License     :  BSD3
--   Maintainer  :  pxqr.sta@gmail.com
--   Stability   :  experimental
--   Portability :  portable
--
--   An efficient implementation of dense integer sets based on
--   Big-Endian PATRICIA trees with buddy suffix compression.
--
--   References:
--
--     * Fast Mergeable Integer Maps (1998) by Chris Okasaki, Andrew Gill
--       <http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.37.5452>
--
--   This implementation performs espessially well then set contains
--   long integer invervals like @[0..2047]@ that are just merged into
--   one interval description. This allow to perform many operations
--   in constant time and space. However if set contain sparse
--   integers like @[1,12,7908,234,897]@ the same operations will take
--   /O(min(W, n))/ which is good enough in most cases.
--
--   Conventions in complexity notation:
--
--     * n — number of elements in a set;
--
--     * W — number bits in a 'Key'. This is 32 or 64 at 32 and 64 bit
--     platforms respectively;
--
--     * O(n) or O(k) — means this operation have complexity O(n) in
--     worst case (e.g. sparse set) or O(k) in best case (e.g. one
--     single interval).
--
--   Note that some operations will take centuries to compute. For
--   example @map id universe@ will a long time to end as well as
--   'filter' applied to 'universe', 'naturals', 'positives' or
--   'negatives'.
--
--   Also note that some operations like 'union', 'intersection' and
--   'difference' have overriden from default fixity, so use these
--   operations with infix syntax carefully.
--
{-# LANGUAGE CPP #-}

#if __GLASGOW_HASKELL__ >= 720
{-# LANGUAGE Safe #-}
#endif

module Data.IntervalSet
       (
         -- * Types
         IntSet(..), Key

         -- * Query
         -- ** Cardinality
       , SB.null
       , size

         -- ** Membership
       , member, notMember

         -- ** Inclusion
       , isSubsetOf, isSupersetOf
--       , isProperSubsetOf, isProperSupersetOf

         -- * Construction
       , empty
       , singleton
       , interval
{-
       , naturals
       , negatives
       , universe
-}
         -- * Modification
       , insert
       , delete

         -- * Map Fold Filter
       , SB.map
       , SB.foldr
       , SB.filter

         -- * Splits
       , split, splitGT, splitLT
       , partition

         -- * Min/Max
       , findMin, findMax

         -- * Combine
       , union, unions
       , intersection, intersections
       , difference, symDiff

         -- ** Monoids
       , Union, Intersection, Difference

         -- * Conversion
         -- *** Arbitary
       , elems
       , toList, fromList

         -- *** Ordered
       , toAscList, toDescList
       , fromAscList

#if defined (TESTING)
         -- * Debug
       , isValid, splitFin
#endif
       ) where

import Data.IntervalSet.Internal as SB

src/Data/IntSet/Buddy.hs
-- |
--   Copyright   :  (c) Sam T. 2013
--   License     :  BSD3
--   Maintainer  :  pxqr.sta@gmail.com
--   Stability   :  experimental
--   Portability :  portable
--
--   An efficient implementation of dense integer sets based on
--   Big-Endian PATRICIA trees with buddy suffix compression.
--
--   References:
--
--     * Fast Mergeable Integer Maps (1998) by Chris Okasaki, Andrew Gill
--       <http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.37.5452>
--
--   This implementation is pretty similar to Data.IntSet from
--   containers package. It's expected that Data.IntSet.Buddy will be
--   slightly slower in randomized settings (e.g. fromList
--   [1,3..1000000]) but much faster when we have long sequences
--   (e.g. [1..100000]) up to constant time at many main operations
--   not depending on a set size.
--
--   Conventions in complexity notation:
--
--     * n - number of elements in a set;
--
--     * W - number bits in a 'Key'. This is 32 at 32 bit platforms
--     and 64 at 64 bit platforms;
--
--     * O(n) or O(k) means this operation have complexity O(n) in
--     worst case (e.g. sparse set) or O(k) in best case (e.g. one
--     single interval).
--
--   Note that some operations will take centuries to compute. For
--   exsample @map id universe@ will never end as well as 'filter'
--   applied to 'universe', 'naturals', 'positives' or 'negatives'.
--
--   Also note that some operations like 'union', 'intersection' and
--   'difference' have overriden from default fixity, so use these
--   operations with infix syntax carefully.
--
{-# LANGUAGE CPP #-}

#if __GLASGOW_HASKELL__ >= 720
{-# LANGUAGE Safe #-}
#endif

module Data.IntSet.Buddy
       (
         -- * Types
         IntSet(..), Key

         -- * Query
       , SB.null
       , size
       , member, notMember

         -- * Construction
       , empty
       , singleton
       , interval

       , naturals
       , negatives
       , universe

         -- * Modification
       , insert
       , delete

         -- * Map/Fold/Filter
       , SB.map
       , SB.foldr
       , SB.filter

         -- * Splits
       , split, splitGT, splitLT

         -- * Min/Max
       , findMin, findMax

         -- * Combine
       , union, unions
       , intersection, intersections
       , difference

         -- * Conversion
         -- ** Lists
       , elems
       , toList, fromList

#if defined (TESTING)
         -- * Debug
       , isValid, splitFin
#endif
       ) where

import Data.IntSet.Buddy.Internal as SB

bench/Main.hs
-- |
--   Copyright   :  (c) Sam T. 2013
--   License     :  BSD3
--   Maintainer  :  pxqr.sta@gmail.com
--   Stability   :  experimental
--   Portability :  portable
--
--   Each function should be benchmarked at least in the following modes:
--
--     * sparse — to see worst case performance. Taking into account
--     both implementations I think [0,64..N] is pretty sparse.
--
--     * dense — to see expected performance. Again [0,2..N] is pretty
--     dense but not interval yet.
--
--     * interval — to see best case performance. Set should be one
--     single interval like [0..N].
--
--   This should help us unify benchmarks and make it more infomative.
--
{-# LANGUAGE BangPatterns #-}
module Main (main) where

import Criterion.Main

import Data.Bits
import           Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Data.IntSet as S
import Data.IntervalSet as SB
import Data.IntervalSet.ByteString as SB
import Data.List as L
import Data.Word




fromByteString :: ByteString -> S.IntSet
fromByteString = S.fromDistinctAscList . indices 0 . B.unpack
  where
    indices _ []       = []
    indices n (w : ws) = wordIxs n w ++ indices (n + 8) ws

    wordIxs n w = L.map ((n +) . fst) $ L.filter snd $ zip [0..] (bits w)

    bits i = L.map (testBit i) [0..bitSize (0 :: Word8) - 1]

main :: IO ()
main = defaultMain $
  [ bench "fromList/O-2.5K"  $ nf S.fromList [0..2500]
  , bench "fromList/O-5K"  $ nf S.fromList [0..5000]
  , bench "fromList/O-10K" $ nf S.fromList [0..10000]
  , bench "fromList/O-20K" $ nf S.fromList [0..20000]
  , bench "fromList/O-20K" $ nf S.fromList (L.map (* 10) [0..20000])
  , bench "fromList/S-2.5K"  $ nf SB.fromList [0..2500]
  , bench "fromList/S-5K"  $ nf SB.fromList [0..5000]
  , bench "fromList/S-10K" $ nf SB.fromList [0..10000]
  , bench "fromList/S-20K" $ nf SB.fromList [0..20000]
  , bench "fromList/S-20K-sparse" $ nf SB.fromList (L.map (* 10) [0..20000])

  , bench "interval/O-200K" $ whnf S.fromDistinctAscList [0..200000]
  , bench "interval/S-2M"   $ whnf (SB.interval 0) 2000000

  , let !s = S.fromList [1..50000] in
    bench "toList/50K" $ nf S.toList s

  , let !s = SB.fromList [1..50000] in
    bench "toList/50K" $ nf SB.toList s

  , let !bs = B.replicate 10000 255 in
    bench "bitmap/from-10K-O" $ whnf Main.fromByteString bs

  , let !bs = B.replicate (2 ^ (17 :: Int)) 0 in
    bench "bitmap/from-1M-S-empty" $ whnf SB.fromByteString bs

  , let !bs = B.replicate (2 ^ (17 :: Int)) 1 in
    bench "bitmap/from-1M-S-dense" $ whnf SB.fromByteString bs

  , let !bs = B.replicate (2 ^ (17 :: Int)) 255 in
    bench "bitmap/from-1M-S-buddy" $ whnf SB.fromByteString bs

  , let !s = SB.fromList [0, 2 .. 1000000 * 2] in
    bench "bitmap/to-strict-1M-S-dense" $ whnf SB.toByteString s

  , let !s = interval 0 10000000 in
    bench "bitmap/to-strict-10M-S-buddy" $ whnf SB.toByteString s

  , let !s = S.fromList [0..1000000] in
    bench "member/1M" $ nf (L.all (`S.member` s)) [50000..100000]

  , let !s = SB.fromList [0..1000000] in
    bench "member/1M" $ nf (L.all (`SB.member` s)) [50000..100000]
  ]
  ++ concat
  [ splitBenchs
  , splitGTBenchs
  , splitLTBenchs

  , partBenchs

  , mergeTempl S.union        SB.union        "union"
  , mergeTempl S.intersection SB.intersection "intersection"
  , mergeTempl S.difference   SB.difference   "difference"
  , mergeTempl   symDiff'     SB.symDiff      "symmetric-difference"
  ]

symDiff' :: S.IntSet -> S.IntSet -> S.IntSet
symDiff' a b = (a `S.union` b) `S.difference` (a `S.intersection` b)

partBenchs :: [Benchmark]
partBenchs = complexBench "partition" 10000 (snd . S.partition even)
             (snd . SB.partition even)

splitBenchs :: [Benchmark]
splitBenchs = complexBench "split" 1000000 (chunk S.split) (chunk SB.split)
  where
    chunk op s = L.foldr ((snd .) . op) s points
    points     = [100,200..1000000]

splitGTBenchs :: [Benchmark]
splitGTBenchs = complexBench "split/GT" 1000000 (chunk S.split) (chunkGT SB.splitGT)
  where
    chunkGT op s = L.foldr op s points
    chunk op s = L.foldr ((snd .) . op) s points
    points     = [100,200..1000000]

splitLTBenchs :: [Benchmark]
splitLTBenchs = complexBench "split/LT" 1000000 (chunk S.split) (chunkGT SB.splitLT)
  where
    chunkGT op s = L.foldr op s points
    chunk op s = L.foldr ((fst .) . op) s points
    points     = reverse [100,200..1000000]

{--------------------------------------------------------------------
  Benchmark Templates
--------------------------------------------------------------------}

type Name = String

type Template =   Name  -- name of benchmark
              ->  Int   -- size of int set
              -> (S.IntSet -> S.IntSet)
              -> (SB.IntSet -> SB.IntSet)
              -> [Benchmark]

type Gen a = Int -> a

genericBench :: Name -> Gen S.IntSet -> Gen SB.IntSet -> Template
genericBench _type genA genB name n f g =
  [ let !s = genA n in bench (name ++ "/O-" ++ show n ++ "-" ++ _type) $ whnf f s
  , let !s = genB n in bench (name ++ "/S-" ++ show n ++ "-" ++ _type) $ whnf g s
  ]

sparseSB :: Gen SB.IntSet
sparseSB n = SB.fromList [0, 64..n * 64 ]

sparseS :: Gen S.IntSet
sparseS n = S.fromDistinctAscList [0, 64..n * 64 ]

denseSB :: Gen SB.IntSet
denseSB n = SB.fromList [0, 2 .. n * 2]

denseS :: Gen S.IntSet
denseS n = S.fromDistinctAscList [0, 2 .. n * 2]

intervalS :: Gen S.IntSet
intervalS n = S.fromDistinctAscList [0..n]

intervalSB :: Gen SB.IntSet
intervalSB n = SB.fromList [0..n]

sparseBench :: Template
sparseBench = genericBench "sparse" sparseS sparseSB

denseBench :: Template
denseBench = genericBench "dense" denseS denseSB

intervalBench :: Template
intervalBench = genericBench "interval" intervalS intervalSB

complexBench :: Template
complexBench name n f g = L.concatMap (\t -> t name n f g) templs
  where
    templs = [sparseBench, denseBench, intervalBench]


mergeTempl :: (S.IntSet  -> S.IntSet  -> S.IntSet)
           -> (SB.IntSet -> SB.IntSet -> SB.IntSet)
           -> String -> [Benchmark]
mergeTempl sop bop n =
  [ let (!a, !b) = (S.fromList [0,64..10000 * 64], S.fromList [1,65..10000 * 64]) in
    bench (n ++"/O-10K-sparse-disjoint")  $ whnf (uncurry sop) (a, b)

  , let (!a, !b) = (S.fromList [0,64..10000 * 64], S.fromList [0,64..10000 * 64]) in
    bench (n ++"/O-10K-sparse-overlap")  $ whnf (uncurry sop) (a, b)

  , let (!a, !b) = (SB.fromList [0,64..10000 * 64], SB.fromList [1,65..10000 * 64]) in
    bench (n ++ "/S-10K-sparse-disjoint") $ whnf (uncurry bop) (a, b)

  , let (!a, !b) = (SB.fromList [0,64..10000 * 64], SB.fromList [0,64..10000 * 64]) in
     bench (n ++ "/S-10K-sparse-overlap") $ whnf (uncurry bop) (a, b)

  , let (!a, !b) = (S.fromList [0,2..500000 * 2], S.fromList [1,3..500000 * 2]) in
    bench (n ++ "/O-500K-dense-disjoint")  $ whnf (uncurry sop) (a, b)

  , let (!a, !b) = (S.fromList [0,2..500000 * 2], S.fromList [0,2..500000 * 2]) in
    bench (n ++ "/O-500K-dense-overlap")  $ whnf (uncurry sop) (a, b)

  , let (!a, !b) = (SB.fromList [0,2..500000 * 2], SB.fromList [1,3..500000 * 2]) in
    bench (n ++ "/S-500K-dense-disjoint") $ whnf (uncurry bop) (a, b)

  , let (!a, !b) = (SB.fromList [0,2..500000 * 2], SB.fromList [0,2..500000 * 2]) in
    bench (n ++ "/S-500K-dense-overlap") $ whnf (uncurry bop) (a, b)

  , let (!a, !b) = (S.fromList [0..500000], S.fromList [0..500000]) in
    bench (n ++ "/O-500K-buddy")  $ whnf (uncurry sop) (a, b)

  , let (!a, !b) = (SB.fromList [0..500000], SB.fromList [0..500000]) in
    bench (n ++ "/S-500K-buddy") $ whnf (uncurry bop) (a, b)
  ]